src/brain/index_selection.cpp

```diff
@@ -0,0 +1,449 @@
+  // m configurations
+  for (auto index_pair : result_index_config) {
+    top_indexes.Merge(index_pair.first);
+  }
+}
+
+void IndexSelection::GetAdmissibleIndexes(parser::SQLStatement *query,
+                                          IndexConfiguration &indexes) {
+  // Find out the indexable columns of the given workload.
+  // The following rules define what indexable columns are:
+  // 1. A column that appears in the WHERE clause with format
+  //    ==> Column OP Expr <==
+  //    OP such as {=, <, >, <=, >=, LIKE, etc.}
+  //    Column is a table column name.
+  // 2. GROUP BY (if present)
+  // 3. ORDER BY (if present)
+  // 4. all updated columns for UPDATE query.
+
+  union {
+    parser::SelectStatement *select_stmt;
+    parser::UpdateStatement *update_stmt;
+    parser::DeleteStatement *delete_stmt;
+    parser::InsertStatement *insert_stmt;
+  } sql_statement;
+
+  switch (query->GetType()) {
+    case StatementType::INSERT:
+      sql_statement.insert_stmt =
+          dynamic_cast<parser::InsertStatement *>(query);
+      // If the insert is along with a select statement, i.e another table's
+      // select output is fed into this table.
+      if (sql_statement.insert_stmt->select != nullptr) {
+        IndexColsParseWhereHelper(
+            sql_statement.insert_stmt->select->where_clause.get(), indexes);
+      }
+      break;
+
+    case StatementType::DELETE:
+      sql_statement.delete_stmt =
+          dynamic_cast<parser::DeleteStatement *>(query);
+      IndexColsParseWhereHelper(sql_statement.delete_stmt->expr.get(), indexes);
+      break;
+
+    case StatementType::UPDATE:
+      sql_statement.update_stmt =
+          dynamic_cast<parser::UpdateStatement *>(query);
+      IndexColsParseWhereHelper(sql_statement.update_stmt->where.get(),
+                                indexes);
+      break;
+
+    case StatementType::SELECT:
+      sql_statement.select_stmt =
+          dynamic_cast<parser::SelectStatement *>(query);
+      IndexColsParseWhereHelper(sql_statement.select_stmt->where_clause.get(),
+                                indexes);
+      IndexColsParseOrderByHelper(sql_statement.select_stmt->order, indexes);
+      IndexColsParseGroupByHelper(sql_statement.select_stmt->group_by, indexes);
+      break;
+
+    default:
+      LOG_ERROR("Cannot handle DDL statements");
+      PELOTON_ASSERT(false);
+  }
+}
+
+void IndexSelection::IndexColsParseWhereHelper(
+    const expression::AbstractExpression *where_expr,
+    IndexConfiguration &config) {
+  if (where_expr == nullptr) {
+    LOG_DEBUG("No Where Clause Found");
+    return;
+  }
+  auto expr_type = where_expr->GetExpressionType();
+  const expression::AbstractExpression *left_child;
+  const expression::AbstractExpression *right_child;
+  const expression::TupleValueExpression *tuple_child;
+
+  switch (expr_type) {
+    case ExpressionType::COMPARE_EQUAL:
+      PELOTON_FALLTHROUGH;
```

Comment:
Nit: I don't know if we even enable fallthrough checks in our build, but it seems like in other cases we only use this compiler keyword when there is a body of code after the tag (i.e. case ...: ...; PELOTON_FALLTHROUGH case:...)

I am relatively certain that if the body is empty people just leave out the fallthrough?